src/lib.rs

```diff
@@ -263,6 +264,505 @@ macro_rules! cfg_global_asm {
     };
 }
 
+macro_rules! trap_asm {
+    ($call:literal, $($rest:tt)*) => {
+        asm!(concat!("
+addi sp, sp, -40*4
+
+sw ra, 0*4(sp)
+sw t0, 1*4(sp)
+sw t1, 2*4(sp)
+sw t2, 3*4(sp)
+sw t3, 4*4(sp)
+sw t4, 5*4(sp)
+sw t5, 6*4(sp)
+sw t6, 7*4(sp)
+sw a0, 8*4(sp)
+sw a1, 9*4(sp)
+sw a2, 10*4(sp)
+sw a3, 11*4(sp)
+sw a4, 12*4(sp)
+sw a5, 13*4(sp)
+sw a6, 14*4(sp)
+sw a7, 15*4(sp)
+sw s0, 16*4(sp)
+sw s1, 17*4(sp)
+sw s2, 18*4(sp)
+sw s3, 19*4(sp)
+sw s4, 20*4(sp)
+sw s5, 21*4(sp)
+sw s6, 22*4(sp)
+sw s7, 23*4(sp)
+sw s8, 24*4(sp)
+sw s9, 25*4(sp)
+sw s10, 26*4(sp)
+sw s11, 27*4(sp)
+sw gp, 28*4(sp)
+sw tp, 29*4(sp)
+csrrs t1, mepc, x0
+sw t1, 31*4(sp)
+csrrs t1, mstatus, x0
+sw t1, 32*4(sp)
+csrrs t1, mcause, x0
+sw t1, 33*4(sp)
+csrrs t1, mtval, x0
+sw t1, 34*4(sp)
+
+addi s0, sp, 40*4
+sw s0, 30*4(sp)
+
+add a0, sp, zero
+        ",
+            $call,
+        "
+lw t1, 31*4(sp)
+csrrw x0, mepc, t1
+
+lw t1, 32*4(sp)
+csrrw x0, mstatus, t1
+
+lw ra, 0*4(sp)
+lw t0, 1*4(sp)
+lw t1, 2*4(sp)
+lw t2, 3*4(sp)
+lw t3, 4*4(sp)
+lw t4, 5*4(sp)
+lw t5, 6*4(sp)
+lw t6, 7*4(sp)
+lw a0, 8*4(sp)
+lw a1, 9*4(sp)
+lw a2, 10*4(sp)
+lw a3, 11*4(sp)
+lw a4, 12*4(sp)
+lw a5, 13*4(sp)
+lw a6, 14*4(sp)
+lw a7, 15*4(sp)
+lw s0, 16*4(sp)
+lw s1, 17*4(sp)
+lw s2, 18*4(sp)
+lw s3, 19*4(sp)
+lw s4, 20*4(sp)
+lw s5, 21*4(sp)
+lw s6, 22*4(sp)
+lw s7, 23*4(sp)
+lw s8, 24*4(sp)
+lw s9, 25*4(sp)
+lw s10, 26*4(sp)
+lw s11, 27*4(sp)
+lw gp, 28*4(sp)
+lw tp, 29*4(sp)
+lw sp, 30*4(sp)
+
+# SP was restored from the original SP
+mret
+"),
+    $($rest)*)
+    };
+}
+
+#[naked]
+#[no_mangle]
+#[link_section = ".trap"]
+unsafe extern "C" fn default_start_trap1() {
+    extern "C" {
+        fn interrupt1(tf: &mut TrapFrame);
+    }
+    trap_asm!("call ra, {irq}",
```

Comment:
Oh, nice: thanks for the worked example! I had the thought of "maybe we save a _little_ off to the stack, and then call something to save the rest" but it's great to see it in action!

There's a definite speed-vs-size tradeoff here: not just with the "how do we spill the registers to the stack frame" question but with the whole "vectored vs (pseudo)-direct" thing. Rather than picking a particular answer, I'm really curious if there isn't some kind of "pay for what you use" strategy here: LLVM is really good at register allocation and spills, and especially if we're able to push the vectoring all the way through to the user function it sure would be nice to have a "spill-as-you-go" model that ~marks all the registers as "tained" (or whatever the term is). We've got a couple handlers that just copy bytes from the UART fifos to a vec (which, most of the time is `lw; sw` x2 + `inc`), and it sure would be nice to avoid saving 31 other registers to get to use a single `t0` haha.

If that can work, then I think we get to remove the _start_trap step all together, or maybe limit it to cases that require the full trap frame for whatever reason.

Reply:
this spill-as-you-go approach sounds really nice if possible, i've been able to get basic scheduling and dispatching of tasks in RTIC down to something like 120 assembly instructions, meaning the context dump/restore is actually responsible for most of the latency at this point

Reply:
well, the good news is that the existing support in LLVM works a treat; the bad news is that it's not (yet) exposed through any nightlies of rustc. I've got a PR to add it here: https://github.com/rust-lang/rust/pull/111891